text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,119 @@
+```
+<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
+
+<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
+
+<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
+
+<immutable-data-uri>  := <cid> <query-fragment>
+
+<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
+
+<path-query-fragment> := ['/' <path>] <query-fragment>
+
+<query-fragment>      := ['?' <query>] ['#' <fragment>]
+```
+
+Where:
+
+- `<cid>`: follows the [CID format](https://github.com/ipld/cid) which self-describes and encodes:
+    - the base encoding for the string, it's here proposed to use [z-base32 encoding](http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt) for the reasons explained above,
+    - the version of CID for upgradability purposes (v1 now),
+    - the content type or codec of the data being referenced,
+    - the XOR address of the content being referenced
+- `<content-version>`: for future implementation to reference versionable content, using a single address with different versions
+
+- `<type-tag>`: the type tag value if the CID is referencing a MutableData. In the absence of this value, the CID will be assumed to be targetting an ImmutableData
+
+- `<path>`: the path of the file if the CID is referencing a MutableData, which can be accessed through the NFS/Files convention (or other emulations/conventions in the future) and resolve it
+
+- `<query>`: query arguments, to be used by the client app and not for retrieving the content
+
+- `<fragment>`: fragment of the content, to be used by the client app and not for retrieving the content
+
+The resolver function shall first attempt to decode the URL assuming it contains a `<cid>` part. If that step fails (either because it couldn’t decode the CID, or because there was no data at the decoded XOR address), it will do a fallback to assume it’s a public-name URL and it shall try to find the pulic-name MD (with the `sha3` hash of the string), just like any public-name URLs are resolved now by the resolver function.
+
+Note that it's proposed that the resolver should try to decode the CID and attempt to fetch the content before falling back to assume it's a public-name URL, rather than only decoding. The reason behind this is to not eclipse a public-name which happens to be a valid CID string. In such cases it could still be eclipsed if there exist content at the location encoded by the CID string, but at least not do it up front when there is no content stored at such location.
+
+The following are examples of what would become valid XOR-URLs as per this proposal:
+
+ImmutableData XOR-URL:
+`safe://hygjdkfty6m7ag3bckq7eqgeizbtjk915c3jbrcgtisad8iikbk4xws4jbpky`
+
+MutableData XOR-URL:
+`safe://hyfktce8j75yhmj1dbi1xw5wnb4m3zdydr7wpbzf1a16hc3sbxzu8a9hiqw:15000`
```

Comment:
If we have a new data type which has a XOR name and tag type which behaves different to a MudatbleData, for example, consider the StructuredData is being revived and supported. How can we represent the same as a XOR url?
If my understanding is right it will also have a similar URL as a MutableData would have.

Potential options:
- The CID generation takes [registered codec](https://github.com/ipld/js-cid#new-cidversion-codec-multihash) as an input. But not sure whether custom codec DataTypes will be accepted as a part of the codec list to multicodec and cids repositories. Other options is to fork the repository which comes with maintenance overhead.
- Have different hashing techniques to differentiate Datatypes. When the URL is decoded the `webFetch` can do a switch based on the hashing techinque to resolve the corresponding datatype.

Reply:
Yes this is a good point, and I also thought a bit about it, and I was thinking more towards your first option. If the target content is an immutable data then the codec will likely be the one corresponding to the file itself, but if it's another type of 'structured' data like MutableData, then the codec can be our own custom set of types, which I think it should be possible since I see there are already a few custom ones in the table, like the ones for Ethereum: https://github.com/multiformats/multicodec/blob/master/table.csv#L416